### PR TITLE
End to end tests for Tomcat operator

### PR DESCRIPTION
**Pull request by adam-sandor**
These tests are meant to be the blueprint for future end-to-end tests. This PR also includes significant fixes to the TomcatOperator to fix issues revealed by the tests.

The important characteristics of this test are:
- Builds and deploys the operator to a real Kubernetes cluster
- Uses Github pipelines to set up a fresh Kubernetes cluster and deploy CRDs
- The verification stage is written in JUnit
- Each verification retries the assertions until they succeed or a timout is reached. This allows us to work with high timeouts as success is detected immediately (compared to sleep which waits in both cases).

### CODE REVIEW

**Comment by adam-sandor** [comment]
Hold the reviews, looks like the test is failing.

**Comment by adam-sandor** [comment]
Ready for review

File: .github/workflows/TomcatIntegrationTest.yml [hunk 1]
```diff
@@ -6,35 +8,72 @@ on:
...
         - name: Set up Java and Maven
-          uses: actions/setup-java@v1
+          uses: actions/setup-java@v2
           with:
-            # java-version: ${{ matrix.java }}
             java-version: 15
-        - uses: actions/cache@v2
+            distribution: adopt-hotspot
+
+        - name: cache
```
Comment (metacosm) [inline .github/workflows/TomcatIntegrationTest.yml:41]: This isn't needed anymore with `setup-java` v2. See https://github.com/java-operator-sdk/java-operator-sdk/pull/527 for details. Then again, maybe you need more control on the caching that what's provided by default?

File: .github/workflows/TomcatIntegrationTest.yml [hunk 2]
```diff
@@ -6,35 +8,72 @@ on:
...
+          uses: actions/cache@v2
+          if: ${{ !env.ACT }}
           with:
             path: ~/.m2/repository
             key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
             restore-keys: |
-              ${{ runner.os }}-maven-
+              ${{ runner.os }}-maven-m2
+
+        - name: Set up Maven
```
Comment (metacosm) [inline .github/workflows/TomcatIntegrationTest.yml:50]: Isn't maven set up as part of `setup-java` already?

File: .github/workflows/TomcatIntegrationTest.yml [hunk 3]
```diff
@@ -6,35 +8,72 @@ on:
...
+          run: |
+            for DIMG in "$KIND_CL_NAME-control-plane "; do
+              docker stop $DIMG ; docker rm $DIMG ;
+            done ;
+            sleep 1
+
+        - name: Create Kubernetes KinD Cluster
+          uses: container-tools/kind-action@v1.5.0
+          with:
+            cluster_name: tomcat-integration-test
```
Comment (metacosm) [inline .github/workflows/TomcatIntegrationTest.yml:29]: Shouldn't this use `${KIND_CL_NAME}` instead?

File: tomcat/pom.xml [hunk 4]
```diff
@@ -19,14 +19,14 @@
...
-        <jib-maven-plugin.version>2.7.1</jib-maven-plugin.version>
+        <jib-maven-plugin.version>3.1.4</jib-maven-plugin.version>
     </properties>
 
     <dependencies>
         <dependency>
             <groupId>io.javaoperatorsdk</groupId>
             <artifactId>operator-framework</artifactId>
-            <version>1.8.4</version>
+            <version>1.9.2</version>
```
Comment (metacosm) [inline tomcat/pom.xml:29]: 1.9.6 is available 😉

File: tomcat/src/test/java/io/javaoperatorsdk/operator/sample/IntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+public class IntegrationTest {
+
+    final static String TEST_NS = "tomcat-test";
+
+    final static Logger log = LoggerFactory.getLogger(IntegrationTest.class);
+
+    @Test
+    public void test() {
+        Config config = new ConfigBuilder().withNamespace(null).build();
```
Comment (metacosm) [inline tomcat/src/test/java/io/javaoperatorsdk/operator/sample/IntegrationTest.java:31]: You could probably use: `Config.autoConfigure(null)` instead here.

**Review by metacosm — COMMENTED** [review]
Sorry for the late review… 😓 
I feel that we can improve all of this somehow. Maybe by making the `IntegrationTestSupport` class found in JOSDK more generic?

**Comment by adam-sandor** [comment]
Thanks @metacosm I'll fix these with another PR.

File: tomcat/src/test/java/io/javaoperatorsdk/operator/sample/IntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+
+import static java.util.concurrent.TimeUnit.*;
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.notNullValue;
+
+public class IntegrationTest {
```
Comment (csviri) [inline tomcat/src/test/java/io/javaoperatorsdk/operator/sample/IntegrationTest.java:23]: Could we brake this method into smaller chuns based on semantics, it's kinda hard to follow this way.

**Review by csviri — COMMENTED** [review]
LGTM
